examples/federation-axum/instance.rs

```diff
@@ -0,0 +1,170 @@
+    UrlVerifier,
+    APUB_JSON_CONTENT_TYPE,
+};
+
+use activitypub_federation::core::axum::{verify_request_payload, DigestVerified};
+use async_trait::async_trait;
+use axum::{
+    body,
+    body::{Body, BoxBody},
+    extract::{Json, OriginalUri, State},
+    middleware,
+    response::IntoResponse,
+    routing::{get, post},
+    Extension,
+    Router,
+};
+use http::{header::CONTENT_TYPE, HeaderMap, Request, Response};
+use reqwest::Client;
+use std::{
+    net::SocketAddr,
+    str::FromStr,
+    sync::{Arc, Mutex},
+};
+use tokio::task;
+use tower::ServiceBuilder;
+use tower_http::ServiceBuilderExt;
+use url::Url;
+
+pub type InstanceHandle = Arc<Instance>;
+
+pub struct Instance {
+    /// This holds all library data
+    local_instance: LocalInstance,
+    /// Our "database" which contains all known users (local and federated)
+    pub users: Mutex<Vec<MyUser>>,
+    /// Same, but for posts
+    pub posts: Mutex<Vec<MyPost>>,
+}
+
+/// Use this to store your federation blocklist, or a database connection needed to retrieve it.
+#[derive(Clone)]
+struct MyUrlVerifier();
+
+#[async_trait]
+impl UrlVerifier for MyUrlVerifier {
+    async fn verify(&self, url: &Url) -> Result<(), &'static str> {
+        if url.domain() == Some("malicious.com") {
+            Err("malicious domain")
+        } else {
+            Ok(())
+        }
+    }
+}
+
+impl Instance {
+    pub fn new(hostname: String) -> Result<InstanceHandle, Error> {
+        let settings = InstanceSettings::builder()
+            .debug(true)
+            .url_verifier(Box::new(MyUrlVerifier()))
+            .build()?;
+        let local_instance =
+            LocalInstance::new(hostname.clone(), Client::default().into(), settings);
+        let local_user = MyUser::new(generate_object_id(&hostname)?, generate_actor_keypair()?);
+        let instance = Arc::new(Instance {
+            local_instance,
+            users: Mutex::new(vec![local_user]),
+            posts: Mutex::new(vec![]),
+        });
+        Ok(instance)
+    }
+
+    pub fn local_user(&self) -> MyUser {
+        self.users.lock().unwrap().first().cloned().unwrap()
+    }
+
+    pub fn local_instance(&self) -> &LocalInstance {
+        &self.local_instance
+    }
+
+    pub fn listen(instance: &InstanceHandle) -> Result<(), Error> {
+        let hostname = instance.local_instance.hostname();
+        let instance = instance.clone();
+        let app = Router::new()
+            .route("/inbox", post(http_post_user_inbox))
+            .layer(
+                ServiceBuilder::new()
+                    .map_request_body(body::boxed)
+                    .layer(middleware::from_fn(verify_request_payload)),
+            )
+            .route("/objects/:user_name", get(http_get_user))
+            .with_state(instance);
+
+        // run it
+        let addr = SocketAddr::from_str(hostname)?;
+        let server = axum::Server::bind(&addr).serve(app.into_make_service());
+
+        task::spawn(server);
```

Comment:
This probably should have a comment that in reallife scenarios this should not be on a separate spawn thread as otherwise the app would just exit instead of waiting :)

Reply:
It's not exiting, there was some request part missing on the signature verify logic. 
This is fixed now.